kep/numa-nodes-scoring/README.md

```diff
@@ -0,0 +1,297 @@
+    - [Description of substracting resources](#description-of-substracting-resources)
+    - [Description of normalizing score](#description-of-normalizing-score)
+- [Use cases](#use-cases)
+- [Known limitations](#known-limitations)
+- [Test plans](#test-plans)
+- [Graduation criteria](#graduation-criteria)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature enablement and rollback](#feature-enablement-and-rollback)
+- [Implementation history](#implementation-history)
+<!-- /toc -->
+
+# Summary
+
+This document describes behaviour of a new ScoringStrategy `LeastNUMANodes` for NodeResourceTopologyMatch plugin that scores nodes based on how many NUMA nodes are required to run a given pod.
+
+# Motivation
+
+Consuming resources from multiple NUMA nodes can cause significant performance degradation
+in latency-critical execution and high-throughput applications.
+Topology Manager assigns resources from least amount of NUMA nodes but the scheduler is unaware
+of different NUMA topologies. The best case scenario would be to schedule pod on the node that
+can satisfy resource requirements from least amount of NUMA nodes to minimize latency.
+
+## Goals
+
+- Make better scheduling decisions that take NUMA topology into consideration.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a maximum amount of NUMA nodes required to satisfy resource requirements.
+
+# Proposal
+
+A new ScoringStrategy `LeastNUMANodes` would check how many NUMA nodes are required to run a given pod. It will use [CRD][1]
+to get available resources on the worker nodes and identify which topology policy is enabled, the same as other ScoringStrategies.
+
+For now available ScoringStrategies are running only for `single-numa-node` TopologyManager policy.
+`LeastNUMANodes` strategy can score nodes running all TopologyManager policies.
+
+## ScoringStrategy implementation details
+
+Since topology of the node is stored in the CR, kube-scheduler should be subscribed for updates of appropriate CRD type.
+Same mechanism is used by other ScoringStrategies for `NodeResourceTopologyMatch` plugin.
+
+### Description of calculating required NUMA nodes
+
+To calculate how many NUMA nodes are required to run given container/pod the plugin will generate combinations of all NUMA nodes for given bitmask and see if there are enough resources in specific combination.
+It's important to notice the we need to mimic how Topology Manger chooses which bitmask is the most suitable if there are multiple combinations of NUMA nodes that can satisfy the request.
+The bitmask with less bits set or if the same number of bits is set then the mask with more lower-numbered bits set wins out([link](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L143)).
+The [Combination][2] function can used for that since it generate combinations from the lowest values.
+
+```go
```

Comment:
I'd avoid using source code to explain the core idea as source code would evolve quickly as time goes. Would you mind either abstracting the logic in a few bullets or embedding pseudo code only?

I think it may read better to put up an example from this section to "Description of normalizing score". For example, suppose we have 2 candidate nodes after going through Filter phase, and illustrate how the NUMA capacity in each node is like, then explain in each section how the logic is implemented.

Reply:
I have added `Example` section.